### PR TITLE
OpInfo for `nn.functional.dropout2d`, revise sample inputs for `dropout`

### PR DESCRIPTION
**Pull request by krshrimali**
Earlier, we were only testing for inputs with the shape of `(5,)` for `nn.functional.dropout`, but since it's used a lot - I feel it's a good idea to test for a few more shapes including scalars. This PR:

1. Revises sample inputs for `nn.functional.dropout`
2. Adds an OpInfo for `nn.functional.dropout2d`.

A note regarding the documentation:

Looks like `nn.functional.dropout2d` also supports inputs of shape `(H, W)` apart from `(N, C, H, W) / (C, H, W)` but the [documentation](https://pytorch.org/docs/stable/generated/torch.nn.Dropout2d.html#torch.nn.Dropout2d) doesn't mention that (`H, W` case). Should that be revised or am I missing anything here? (Filed an issue here: https://github.com/pytorch/pytorch/issues/67892)

```python
# A 2D tensor is a valid input for Dropout2d
In [11]: tensor = torch.randn((3, 4), device='cpu', dtype=torch.float32)
In [12]: dropout2d = torch.nn.Dropout2d(p=0.5)

In [13]: dropout2d(tensor)
Out[13]:
tensor([[-0.1026, -0.0000, -0.0000, -0.0000],
        [-1.5647,  0.0000, -0.0000, -0.5820],
        [-0.0000, -3.2080,  0.1164, -3.6780]])
```

Issue Tracker: https://github.com/pytorch/pytorch/issues/54261

cc: @mruberry @zou3519

### CODE REVIEW

**Comment by mruberry** [comment]
@saketh-are would you review this?

**Review by saketh-are — APPROVED** [review]
Hey @krshrimali, apologies for the delay on this review -- this looks great to me! Thanks for buffing the `dropout` sample inputs and for adding `dropout2d`'s OpInfo.

**Comment by saketh-are** [comment]
@krshrimali this PR is breaking the OpInfo for `torch.nn.feature_alpha_dropout`, which happens to reuse sample_inputs_dropout for its sample_inputs_func. Would you be interested in looking into this?

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -7182,17 +7182,26 @@ def error_inputs_kthvalue(op_info, device, **kwargs):
...
 
+def sample_inputs_dropout(op_info, device, dtype, requires_grad, *,
+                          train=None, min_input_dim=None, **kwargs):
+    make_arg = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
 
-def sample_inputs_dropout(op_info, device, dtype, requires_grad, **kwargs):
-    make_arg = partial(make_tensor, (S,), device=device, dtype=dtype, requires_grad=requires_grad)
+    if min_input_dim:
+        # Create cases with dim ranging from min_input_dim to min_input_dim + 2 (inclusive)
+        cases = [(S,) * i for i in range(min_input_dim, min_input_dim + 3)]
```
Comment (krshrimali) [inline torch/testing/_internal/common_methods_invocations.py:7191]: This is to generalize cases when the minimum input dimension supported is given (`min_input_dim`). For example, `feature_alpha_dropout` supports minimum 2D inputs, hence, the `cases` generated will be:

```
>>> cases = [(S,) * i for i in range(min_input_dim, min_input_dim + 3)]
>>> cases
[(5, 5), (5, 5, 5), (5, 5, 5, 5)]
```

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -13701,10 +13710,58 @@ def ref_pairwise_distance(input1, input2):
...
+            # Please perform random operations outside of vmap as a workaround
+            DecorateInfo(unittest.expectedFailure, 'TestGradients', "test_forward_mode_AD"),
+            DecorateInfo(unittest.expectedFailure, 'TestGradients', "test_inplace_forward_mode_AD"),
+            # Probably because we have used lambda for the op here
+            # AssertionError: JIT Test does not execute any logic
+            DecorateInfo(unittest.expectedFailure, 'TestJit', 'test_variant_consistency_jit'),),
+        gradcheck_wrapper=wrapper_set_seed,
+        supports_forward_ad=True,
+        supports_out=False,
+        sample_inputs_func=partial(sample_inputs_dropout, train=True, min_input_dim=2),
```
Comment (krshrimali) [inline torch/testing/_internal/common_methods_invocations.py:13757]: `feature_alpha_dropout` with `train=True` only supports floating type inputs, while for `train=False` - it supports inputs of all dtypes. Hence 2 OpInfos.

**Review by krshrimali — COMMENTED** [review]
@saketh-are - Thanks for the ping and taking a look. I've generalized the `sample_inputs_dropout` to be reused for `dropout2d` as well as other dropout variants. It can hopefully be used for `dropout3d` as well.

`feature_alpha_dropout` OpInfo was wrong before (my bad), and I've fixed it. Please see the review comments for reasoning.

Tests pass locally but waiting for the CI. Please let me know if you have any questions or suggestions. Appreciate the patience.

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -13701,10 +13710,58 @@ def ref_pairwise_distance(input1, input2):
...
             wrapper_set_seed(torch.nn.functional.dropout, input, *args, **kwargs, inplace=True)),
+    OpInfo(
+        "nn.functional.dropout2d",
+        op=lambda input, *args, **kwargs:
+            wrapper_set_seed(torch.nn.functional.dropout2d, input, *args, **kwargs),
+        ref=_NOTHING,
+        dtypes=floating_types_and(torch.bfloat16),
+        dtypesIfCUDA=floating_types_and(torch.float16, torch.bfloat16),
+        skips=(
+            # torch.autograd.gradcheck.GradcheckError: While computing batched gradients, got:
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:13721]: Nice xfails and comments here

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -13701,10 +13710,58 @@ def ref_pairwise_distance(input1, input2):
...
+            # Probably because we have used lambda for the op here
+            # AssertionError: JIT Test does not execute any logic
+            DecorateInfo(unittest.skip("Skipped!"), 'TestJit', 'test_variant_consistency_jit'),),
+        gradcheck_wrapper=wrapper_set_seed,
+        supports_forward_ad=True,
+        supports_out=False,
+        sample_inputs_func=partial(sample_inputs_dropout, min_input_dim=2),
+        inplace_variant=lambda input, *args, **kwargs:
+            wrapper_set_seed(torch.nn.functional.dropout2d, input, *args, **kwargs, inplace=True)),
+    # In training mode, feature_alpha_dropout currently doesn't support inputs of complex dtype
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:13735]: Cool note for two OpInfos

**Review by mruberry — APPROVED** [review]
Hey @krshrimali! @saketh-are is out this week so I'll merge this for you

**Pull request closed without merging**